File: web_frontend/frontend/src/api/collaboration.ts
import { requestJson } from './http'

export type CollaborationApprovalMode = 'user_controlled' | 'main_agent_delegated'
export type CollaborationSessionStatus = 'draft' | 'running' | 'completed' | 'failed' | 'cancelled'
export type CollaborationRuntimeStatus =
  | 'waiting_for_workers'
  | 'waiting_for_approval'
  | 'waiting_for_dependency'
  | 'resuming_from_event'
export type CollaborationTaskStatus =
  | 'assigned'
  | 'queued'
  | 'accepted'
  | 'planning'
  | 'working'
  | 'blocked'
  | 'submitted'
  | 'revision_requested'
  | 'completed'
  | 'failed'
  | 'cancelled'

export interface CollaborationAgentView {
  package_id: string
  agent_name: string
  agent_description?: string
  source?: string
  available_as_main: boolean
  available_as_worker: boolean
  status?: string | null
}

export interface CollaborationMessageView {
  message_id: string
  collaboration_id: string
  speaker_type: 'user' | 'main_agent' | 'worker_agent' | 'system' | string
  speaker_package_id?: string | null
  message_kind: string
  content: string
  task_id?: string | null
  event_ref?: string | null
  created_at: string
}

export interface CollaborationTaskView {
  task_id: string
  collaboration_id: string
  parent_task_id?: string | null
  assignee_package_id: string
  assignee_session_id?: string | null
  task_text: string
  depends_on: string[]
  delivery_standard: Record<string, any>
  visible_context: Record<string, any>
  input_artifacts: any[]
  status: CollaborationTaskStatus
  result_summary: string
  result_payload: Record<string, any>
  artifact_refs: any[]
  review_notes: string
  created_at: string
  updated_at: string
}

export interface CollaborationManufacturingRequestView {
  request_id: string
  collaboration_id: string
  agent_name: string
  purpose: string
  status: string
  create_agent_session_id?: string | null
  result_payload?: Record<string, any>
  message?: string | null
  created_at?: string
  updated_at?: string
}

export interface CollaborationSessionView {
  collaboration_id: string
  title: string
  main_agent_package_id: string
  main_agent_package_session_id?: string | null
  approval_mode: CollaborationApprovalMode
  execution_config?: {
    model_profile_id?: string | null
    reasoning_intensity?: number | null
  }
  round_index?: number
  started_at?: string | null
  completed_at?: string | null
  statistics?: CollaborationStatisticsView
  status: CollaborationSessionStatus
  runtime_status?: CollaborationRuntimeStatus | null
  runtime_status_payload?: Record<string, any>
  created_at: string
  updated_at: string
  acceptance_workspace?: {
    resource_mode: 'collaboration'
    collaboration_id: string
    workdir: string
  }
  messages?: CollaborationMessageView[]
  tasks?: CollaborationTaskView[]
  manufacturing_requests?: CollaborationManufacturingRequestView[]
}

export interface CollaborationStatisticsView {
  round_index: number
  wall_duration_ms: number | null
  cumulative_task_duration_ms: number
  task_count: number
  task_status_counts: Record<string, number>
  retry_count: number
  model_usage: {
    totals: {
      call_count: number
      input_tokens: number
      output_tokens: number
      total_tokens: number
      reasoning_tokens: number
      cache_hit_tokens: number
      cache_miss_tokens: number
      cache_hit_ratio: number | null
    }
    by_agent: Array<Record<string, any>>
    by_model: Array<Record<string, any>>
    by_task: Array<Record<string, any>>
  }
}

export const collaborationApi = {
  agents: () => requestJson<{ agents: CollaborationAgentView[] }>('/api/collaboration/agents'),
  sessions: () => requestJson<{ sessions: CollaborationSessionView[] }>('/api/collaboration/sessions'),
  createSession: (payload: {
    title: string
    main_agent_package_id: string
    approval_mode: CollaborationApprovalMode
  }) => requestJson<{ session: CollaborationSessionView }>('/api/collaboration/sessions', {
    method: 'POST',
    body: JSON.stringify(payload),
  }),
  session: (collaborationId: string) =>
    requestJson<{ session: CollaborationSessionView }>(`/api/collaboration/sessions/${encodeURIComponent(collaborationId)}`),
  updateSession: (collaborationId: string, payload: Partial<Pick<CollaborationSessionView, 'title' | 'main_agent_package_id' | 'main_agent_package_session_id' | 'approval_mode' | 'execution_config' | 'status'>>) =>
    requestJson<{ session: CollaborationSessionView }>(`/api/collaboration/sessions/${encodeURIComponent(collaborationId)}`, {
      method: 'PATCH',
      body: JSON.stringify(payload),
    }),
  completeSession: (collaborationId: string, payload: { final_summary: string }) =>
    requestJson<{ session: CollaborationSessionView }>(`/api/collaboration/sessions/${encodeURIComponent(collaborationId)}/complete`, {
      method: 'POST',
      body: JSON.stringify(payload),
    }),
  deleteSession: (collaborationId: string) =>
    requestJson<{
      collaboration_id: string
      deleted: boolean
      main_agent_package_id?: string | null
      main_agent_package_session_id?: string | null
      sessions: CollaborationSessionView[]
    }>(`/api/collaboration/sessions/${encodeURIComponent(collaborationId)}`, {
      method: 'DELETE',
    }),
  addMessage: (collaborationId: string, payload: Partial<CollaborationMessageView>) =>
    requestJson<{ session: CollaborationSessionView }>(`/api/collaboration/sessions/${encodeURIComponent(collaborationId)}/messages`, {
      method: 'POST',
      body: JSON.stringify(payload),
    }),
  mainAgentPrompt: (collaborationId: string, message: string) =>
    requestJson<{
      prompt: string
      runtime_tool_access: {
        extra_allowed_tool_ids: string[]
        excluded_tool_ids: string[]
      }
    }>(`/api/collaboration/sessions/${encodeURIComponent(collaborationId)}/main-agent-prompt`, {
      method: 'POST',
      body: JSON.stringify({ message }),
    }),
  createTask: (collaborationId: string, payload: {
    assignee_package_id: string
    task_text: string
    delivery_standard?: Record<string, any>
    visible_context?: Record<string, any>
    input_artifacts?: any[]
    depends_on?: string[]
  }) => requestJson<{ session: CollaborationSessionView }>(`/api/collaboration/sessions/${encodeURIComponent(collaborationId)}/tasks`, {
    method: 'POST',
    body: JSON.stringify(payload),
  }),
  startTask: (collaborationId: string, taskId: string) =>
    requestJson<{ result: Record<string, any> | null; session: CollaborationSessionView; message?: string | null }>(
      `/api/collaboration/sessions/${encodeURIComponent(collaborationId)}/tasks/${encodeURIComponent(taskId)}/start`,
      { method: 'POST' },
    ),
  cancelTask: (collaborationId: string, taskId: string, payload: { review_notes?: string } = {}) =>
    requestJson<{ session: CollaborationSessionView }>(
      `/api/collaboration/sessions/${encodeURIComponent(collaborationId)}/tasks/${encodeURIComponent(taskId)}/cancel`,
      {
        method: 'POST',
        body: JSON.stringify(payload),
      },
    ),
  retryTask: (collaborationId: string, taskId: string, payload: { retry_guidance?: string } = {}) =>
    requestJson<{ session: CollaborationSessionView; task: CollaborationTaskView; replaced_task_id: string }>(
      `/api/collaboration/sessions/${encodeURIComponent(collaborationId)}/tasks/${encodeURIComponent(taskId)}/retry`,
      {
        method: 'POST',
        body: JSON.stringify(payload),
      },
    ),
  resolveTaskApproval: (collaborationId: string, taskId: string, payload: {
    action: 'approve' | 'deny' | 'revise'
    revision_guidance?: string
  }) =>
    requestJson<{ result: Record<string, any> | null; session: CollaborationSessionView; message?: string | null }>(
      `/api/collaboration/sessions/${encodeURIComponent(collaborationId)}/tasks/${encodeURIComponent(taskId)}/approval`,
      {
        method: 'POST',
        body: JSON.stringify(payload),
      },
    ),
  dispatchReady: (collaborationId: string, limit?: number) =>
    requestJson<{ started_count: number; results: Record<string, any>[]; session: CollaborationSessionView }>(
      `/api/collaboration/sessions/${encodeURIComponent(collaborationId)}/dispatch-ready`,
      {
        method: 'POST',
        body: JSON.stringify(limit ? { limit } : {}),
      },
    ),
  updateTask: (collaborationId: string, taskId: string, payload: Partial<CollaborationTaskView>) =>
    requestJson<{ session: CollaborationSessionView }>(
      `/api/collaboration/sessions/${encodeURIComponent(collaborationId)}/tasks/${encodeURIComponent(taskId)}`,
      {
        method: 'PATCH',
        body: JSON.stringify(payload),
      },
    ),
}

File: web_frontend/backend/routes/collaboration.py
from __future__ import annotations

from dataclasses import asdict
from typing import Any

from fastapi import APIRouter, HTTPException

from agent_factory.collaboration_system import CollaborationService
from agent_factory.collaboration_system.store import CollaborationStoreError, SYSTEM_CHAT_PACKAGE_ID
from agent_factory.factory_graph.frontend_bridge.agent_package_runtime import AgentPackageRuntimeManager
from web_frontend.backend.runtime_bridge import RuntimeBridge


def create_collaboration_router(runtime_bridge: RuntimeBridge, service: CollaborationService) -> APIRouter:
    router = APIRouter(prefix="/api/collaboration")

    @router.get("/agents")
    async def list_collaboration_agents():
        return {"agents": _collaboration_agents(runtime_bridge)}

    @router.get("/sessions")
    async def list_sessions():
        return {"sessions": service.store.list_sessions()}

    @router.post("/sessions")
    async def create_session(payload: dict[str, Any]):
        try:
            session = service.store.create_session(
                title=str(payload.get("title") or ""),
                main_agent_package_id=payload.get("main_agent_package_id"),
                approval_mode=str(payload.get("approval_mode") or ""),
            )
        except Exception as exc:
            raise _http_error(exc) from exc
        return {"session": session}

    @router.get("/sessions/{collaboration_id}")
    async def get_session(collaboration_id: str):
        try:
            return {"session": service.store.get_session(collaboration_id)}
        except Exception as exc:
            raise _http_error(exc) from exc

    @router.patch("/sessions/{collaboration_id}")
    async def update_session(collaboration_id: str, payload: dict[str, Any]):
        try:
            return {"session": service.store.update_session(collaboration_id, payload)}
        except Exception as exc:
            raise _http_error(exc) from exc

    @router.post("/sessions/{collaboration_id}/complete")
    async def complete_session(collaboration_id: str, payload: dict[str, Any]):
        try:
            return {"session": service.complete_session(collaboration_id, {**payload, "speaker_type": "user"})}
        except Exception as exc:
            raise _http_error(exc) from exc

    @router.delete("/sessions/{collaboration_id}")
    async def delete_session(collaboration_id: str):
        try:
            return service.delete_session(collaboration_id)
        except Exception as exc:
            raise _http_error(exc) from exc

    @router.post("/sessions/{collaboration_id}/messages")
    async def add_message(collaboration_id: str, payload: dict[str, Any]):
        try:
            return {"session": service.store.add_message(collaboration_id, payload)}
        except Exception as exc:
            raise _http_error(exc) from exc

    @router.post("/sessions/{collaboration_id}/main-agent-prompt")
    async def main_agent_prompt(collaboration_id: str, payload: dict[str, Any]):
        try:
            prompt = service.build_main_agent_prompt(
                collaboration_id=collaboration_id,
                user_message=str(payload.get("message") or ""),
                worker_agents=[
                    agent for agent in _collaboration_agents(runtime_bridge)
                    if bool(agent.get("available_as_worker"))
                ],
            )
            return {"prompt": prompt}
        except Exception as exc:
            raise _http_error(exc) from exc

    @router.post("/sessions/{collaboration_id}/tasks")
    async def create_task(collaboration_id: str, payload: dict[str, Any]):
        try:
            return {"session": service.create_task(collaboration_id, payload)}
        except Exception as exc:
            raise _http_error(exc) from exc

    @router.post("/sessions/{collaboration_id}/dispatch-ready")
    async def dispatch_ready_tasks(collaboration_id: str, payload: dict[str, Any] | None = None):
        try:
            limit = payload.get("limit") if isinstance(payload, dict) else None
            return service.dispatch_ready(
                collaboration_id,
                limit=int(limit) if limit is not None else None,
            )
        except Exception as exc:
            raise _http_error(exc) from exc

    @router.post("/sessions/{collaboration_id}/tasks/{task_id}/start")
    async def start_task(collaboration_id: str, task_id: str):
        try:
            response = service.start_task(collaboration_id, task_id)
            result = response.get("result")
            return {
                "result": asdict(result) if result is not None else None,
                "session": response["session"],
                "message": response.get("message"),
            }
        except Exception as exc:
            raise _http_error(exc) from exc

    @router.post("/sessions/{collaboration_id}/tasks/{task_id}/cancel")
    async def cancel_task(collaboration_id: str, task_id: str, payload: dict[str, Any] | None = None):
        try:
            return {"session": service.cancel_task(collaboration_id, task_id, payload or {})}
        except Exception as exc:
            raise _http_error(exc) from exc

    @router.post("/sessions/{collaboration_id}/tasks/{task_id}/approval")
    async def resolve_task_approval(collaboration_id: str, task_id: str, payload: dict[str, Any]):
        try:
            response = service.resolve_task_approval(collaboration_id, task_id, payload)
            result = response.get("result")
            return {
                "result": asdict(result) if result is not None else None,
                "session": response["session"],
                "message": response.get("message"),
            }
        except Exception as exc:
            raise _http_error(exc) from exc

    @router.patch("/sessions/{collaboration_id}/tasks/{task_id}")
    async def update_task(collaboration_id: str, task_id: str, payload: dict[str, Any]):
        try:
            return {"session": service.update_task(collaboration_id, task_id, payload)}
        except Exception as exc:
            raise _http_error(exc) from exc

    return router


def _agent_package_runtime(runtime_bridge: RuntimeBridge) -> AgentPackageRuntimeManager:
    adapter = runtime_bridge.adapter
    runtime = getattr(adapter, "agent_package_runtime", None) if adapter is not None else None
    return runtime if isinstance(runtime, AgentPackageRuntimeManager) else AgentPackageRuntimeManager()


def _collaboration_agents(runtime_bridge: RuntimeBridge) -> list[dict[str, Any]]:
    runtime = _agent_package_runtime(runtime_bridge)
    packages = runtime.list_packages()
    agents: list[dict[str, Any]] = [
        {
            "package_id": SYSTEM_CHAT_PACKAGE_ID,
            "agent_name": "闲聊主 Agent",
            "agent_description": "使用系统闲聊作为协作主 Agent。",
            "source": "system",
            "available_as_main": True,
            "available_as_worker": False,
        }
    ]
    for package in packages:
        package_id = str(package.get("package_id") or "").strip()
        if not package_id or package_id == SYSTEM_CHAT_PACKAGE_ID:
            continue
        agents.append(
            {
                "package_id": package_id,
                "agent_name": package.get("agent_name") or package.get("name") or package_id,
                "agent_description": package.get("agent_description") or "",
                "source": "package",
                "available_as_main": True,
                "available_as_worker": True,
                "status": package.get("status"),
            }
        )
    return agents


def _http_error(exc: Exception) -> HTTPException:
    if isinstance(exc, CollaborationStoreError):
        return HTTPException(status_code=400, detail=str(exc))
    if isinstance(exc, FileNotFoundError):
        return HTTPException(status_code=404, detail=str(exc))
    return HTTPException(status_code=500, detail=f"{type(exc).__name__}: {exc}")
